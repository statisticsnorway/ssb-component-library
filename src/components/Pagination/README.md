Pagination
========

> A component for navigating through pages split up in numbers

### Usage

#### HTML

```html
<div class="ssb-pagination">
  <div class="direction-button">
    <i class="chevron-icon">{chevron left icon}</i>Forrige
  </div>
  <div class="nav-button-square selected" onclick="{select this}">1</div>
  <div class="nav-button-square" onclick="{select this}">2</div>
  <div class="nav-button-square" onclick="{select this}">3</div>
  <div class="nav-button-square" onclick="{select this}">4</div>
  <div class="nav-button-square" onclick="{select this}">5</div>
  <div class="nav-button-square" onclick="{select this}">6</div>
  <div class="nav-button-square" onclick="{select this}">7</div>
  <div class="nav-button-square" onclick="{select this}">8</div>
  <div class="dotted-indicator">...</div>
  <div class="nav-button-square" onclick="{select this}">20</div>
  <div class="direction-button">
    Neste<i class="chevron-icon">{chevron right icon}</i>
  </div>
</div>
```

#### React

```jsx harmony
const items = [
  { text: '1', path: '/1' },
  { text: '2', path: '/2' },
  { text: '3', path: '/3' },
  ...
  { text: '20', path: '/20' },
];
<Pagination items={items} selectedPage={items[4]} />
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| items | arraOf({text (string), path (string)}) | Required prop to render the buttons |
| onSelect | func | Callback function |
| selectedPage | object | Pre-selects a page on render |
