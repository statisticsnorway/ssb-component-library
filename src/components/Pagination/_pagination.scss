$direction-button-height: 30px;
$direction-button-width: 80px;

.ssb-pagination {
  @include roboto;
  align-items: center;
  display: flex;
  line-height: 1.25;

  .nav-button-square {
    @include reset-button;
    align-items: center;
    border: 1px solid transparent;
    border-radius: 2px;
    color: $ssb-green-4;
    cursor: pointer;
    display: flex;
    height: 30px;
    justify-content: center;
    margin: 0 4px;
    width: 30px;

    &.selected {
      background: $ssb-green-4;
      color: $ssb-white;
      font-weight: bold;
    }

    &:hover {
      border: 1px solid $ssb-green-4;
      border-radius: 2px;
      font-weight: bold;
    }

    &:focus {
      @include focus-ring;
    }
  }

  .dotted-indicator {
    align-items: center;
    display: flex;
    justify-content: center;
    margin: 0 4px;
    width: 30px;
  }

  .direction-button {
    @include reset-button;
    align-items: center;
    border: 1px solid transparent;
    cursor: pointer;
    display: flex;
    font-size: 16px;
    height: $direction-button-height;
    justify-content: center;
    padding: 4px 10px;

    &:first-of-type {
      text-align: right;
    }

    .chevron-icon {
      color: $ssb-green-4;
      font-size: 16px;
    }

    &:hover {
      border-radius: 2px;
      color: $ssb-green-4;
      font-weight: bold;
      border: 1px solid $ssb-green-4;
    }

    &:focus {
      @include focus-ring;
    }

    &.previous {
      .chevron-icon {
        margin-right: 2px;
      }
    }

    &.next {
      .chevron-icon {
        margin-left: 2px;
      }
    }
  }
}
