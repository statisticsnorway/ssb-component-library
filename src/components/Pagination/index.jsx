import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ChevronLeft, ChevronRight } from 'react-feather';

const Pagination = ({
	className, items, labelNext, labelPrevious, onSelect, selectedPage,
}) => {
	const [selected, setSelected] = useState(selectedPage || items[0]);
	const [currentButtons, updateCurrentButtons] = useState([{}]);
	const [showLeftDots, updateLeftDots] = useState(false);
	const [showRightDots, updateRightDots] = useState(true);

	useEffect(() => {
		const showItems = items;
		if (items.length < 8) {
			updateLeftDots(false);
			updateRightDots(false);
			updateCurrentButtons(items);
		} else if (items.indexOf(selected) < 7) {
			updateLeftDots(false);
			updateCurrentButtons(showItems.slice(0, 8));
			if (items.length > currentButtons.length) {
				updateRightDots(true);
			}
		} else if (items.indexOf(selected) > (items.length - 7)) {
			updateCurrentButtons(showItems.slice((showItems.length - 8), showItems.length));
			updateRightDots(false);
			updateLeftDots(true);
		} else {
			updateCurrentButtons(showItems.slice((showItems.indexOf(selected) - 3), (showItems.indexOf(selected) + 3)));
			updateLeftDots(true);
			updateRightDots(true);
		}
	}, [selected]);

	const handleSelection = sel => {
		setSelected(sel);
		onSelect(sel);
	};

	return (
		<nav className={`ssb-pagination${className ? ` ${className}` : ''}`}>
			<button
				className="direction-button previous"
				onClick={() => handleSelection(items[items.indexOf(selected) - 1])}
				disabled={selected === items[0]}
			>	<ChevronLeft size={16} className="chevron-icon" />
				<span>{labelPrevious}</span>
			</button>
			<button className={`nav-button-square${items[0] === selected ? ' selected' : ''}`} onClick={() => handleSelection(items[0])}>
				{items[0].text}
			</button>
			{showLeftDots && <div className="dotted-indicator">...</div>}
			{currentButtons && currentButtons.map(item => (
				item !== items[0] && item !== items[items.length - 1] && (
					<button
						className={`nav-button-square ${item.text}${item === selected ? ' selected' : ''}`}
						onClick={() => handleSelection(item)}
						key={`${item.path}`}
					>{item.text}
					</button>
				)
			))}
			{showRightDots && <div className="dotted-indicator">...</div>}
			<button
				className={`nav-button-square${items[items.length - 1] === selected ? ' selected' : ''}`}
				onClick={() => handleSelection(items[items.length - 1])}
			>{items[items.length - 1].text}
			</button>
			<button
				className="direction-button next"
				onClick={() => handleSelection(items[items.indexOf(selected) + 1])}
				disabled={selected === items[items.length - 1]}
			>	<span>{labelNext}</span>
				<ChevronRight size={16} className="chevron-icon" />
			</button>
		</nav>
	);
};

Pagination.defaultProps = {
	onSelect: () => {},
	labelNext: 'Next',
	labelPrevious: 'Previous',
};

Pagination.propTypes = {
	className: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		text: PropTypes.string,
		path: PropTypes.string,
	})).isRequired,
	labelNext: PropTypes.string,
	labelPrevious: PropTypes.string,
	onSelect: PropTypes.func,
	selectedPage: PropTypes.object,
};

export default Pagination;
