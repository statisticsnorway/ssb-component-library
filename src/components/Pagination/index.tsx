import React, { useEffect, useState } from 'react'
import { ChevronLeft, ChevronRight } from 'react-feather'
import classNames from '../../utils/utils'

interface Item {
  text: string
  path: string
}
interface PaginationProps {
  className?: string
  items: Item[]
  labelNext?: string
  labelPrevious?: string
  onSelect?: (selected: Item) => void
  selectedPage?: Item
}

const Pagination: React.FC<PaginationProps> = ({
  className,
  items,
  labelNext = 'Next',
  labelPrevious = 'Previous',
  onSelect = () => {},
  selectedPage,
}) => {
  const [selected, setSelected] = useState(selectedPage || items[0])
  const [currentButtons, setCurrentButtons] = useState([{}])
  const [showLeftDots, setShowLeftDots] = useState(false)
  const [showRightDots, setShowRightDots] = useState(true)

  useEffect(() => {
    const showItems = items
    if (items.length < 8) {
      setShowLeftDots(false)
      setShowRightDots(false)
      setCurrentButtons(items)
    } else if (items.indexOf(selected) < 7) {
      setShowLeftDots(false)
      setCurrentButtons(showItems.slice(0, 8))
      if (items.length > currentButtons.length) {
        setShowRightDots(true)
      }
    } else if (items.indexOf(selected) > items.length - 7) {
      setCurrentButtons(showItems.slice(showItems.length - 8, showItems.length))
      setShowRightDots(false)
      setShowLeftDots(true)
    } else {
      setCurrentButtons(showItems.slice(showItems.indexOf(selected) - 3, showItems.indexOf(selected) + 3))
      setShowLeftDots(true)
      setShowRightDots(true)
    }
  }, [selected])

  const handleSelection = (sel: Item) => {
    setSelected(sel)
    onSelect(sel)
  }

  return (
    <nav className={classNames('ssb-pagination', className)}>
      <button
        className='direction-button previous'
        onClick={() => handleSelection(items[items.indexOf(selected) - 1])}
        disabled={selected === items[0]}
      >
        {' '}
        <ChevronLeft size={16} className='chevron-icon' />
        <span>{labelPrevious}</span>
      </button>
      <button
        className={`nav-button-square${items[0] === selected ? ' selected' : ''}`}
        onClick={() => handleSelection(items[0])}
      >
        {items[0].text}
      </button>
      {showLeftDots && <div className='dotted-indicator'>...</div>}
      {currentButtons?.map(
        (item: Item) =>
          item !== items[0] &&
          item !== items[items.length - 1] && (
            <button
              className={`nav-button-square ${item.text}${item === selected ? ' selected' : ''}`}
              onClick={() => handleSelection(item)}
              key={`${item.path}`}
            >
              {item.text}
            </button>
          )
      )}
      {showRightDots && <div className='dotted-indicator'>...</div>}
      <button
        className={`nav-button-square${items[items.length - 1] === selected ? ' selected' : ''}`}
        onClick={() => handleSelection(items[items.length - 1])}
      >
        {items[items.length - 1].text}
      </button>
      <button
        className='direction-button next'
        onClick={() => handleSelection(items[items.indexOf(selected) + 1])}
        disabled={selected === items[items.length - 1]}
      >
        {' '}
        <span>{labelNext}</span>
        <ChevronRight size={16} className='chevron-icon' />
      </button>
    </nav>
  )
}

export default Pagination
