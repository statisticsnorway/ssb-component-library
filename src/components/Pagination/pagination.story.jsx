import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import { ChevronLeft, ChevronRight } from 'react-feather';
import Pagination from './index';

const items = [
	{ text: '1', path: '/1' },
	{ text: '2', path: '/2' },
	{ text: '3', path: '/3' },
	{ text: '4', path: '/4' },
	{ text: '5', path: '/5' },
	{ text: '6', path: '/6' },
	{ text: '7', path: '/7' },
	{ text: '8', path: '/8' },
	{ text: '9', path: '/9' },
	{ text: '10', path: '/10' },
	{	text: '11', path: '/11'	},
	{	text: '12', path: '/12'	},
	{	text: '13', path: '/13'	},
	{	text: '14', path: '/14'	},
	{	text: '15', path: '/15'	},
	{	text: '16', path: '/16'	},
	{	text: '17', path: '/17'	},
	{	text: '18', path: '/18'	},
	{	text: '19', path: '/19'	},
	{	text: '20', path: '/20'	},
];

storiesOf('Pagination', module).addDecorator(centered)
	.add('Default', () => (
		<div>
			<Pagination items={items} selectedPage={items[4]} />
		</div>
	))
	.add('Html', () => (
		<div className="ssb-pagination">
			<div className="direction-button">
				<ChevronLeft className="chevron-icon" size={18} />Forrige
			</div>
			<div className="nav-button-square selected">1</div>
			<div className="nav-button-square">2</div>
			<div className="nav-button-square">3</div>
			<div className="nav-button-square">4</div>
			<div className="nav-button-square">5</div>
			<div className="nav-button-square">6</div>
			<div className="nav-button-square">7</div>
			<div className="nav-button-square">8</div>
			<div className="dotted-indicator">...</div>
			<div className="nav-button-square">20</div>
			<div className="direction-button">
				Neste<ChevronRight className="chevron-icon" size={18} />
				</div>
		</div>
	));
