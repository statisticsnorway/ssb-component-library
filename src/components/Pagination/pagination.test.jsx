import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from './index';

const items = [
	{ text: '1', path: '/1' },
	{ text: '2', path: '/2' },
	{ text: '3', path: '/3' },
	{ text: '4', path: '/4' },
	{ text: '5', path: '/5' },
	{ text: '6', path: '/6' },
	{ text: '7', path: '/7' },
	{ text: '8', path: '/8' },
	{ text: '9', path: '/9' },
	{ text: '10', path: '/10' },
	{	text: '11', path: '/11'	},
	{	text: '12', path: '/12'	},
	{	text: '13', path: '/13'	},
	{	text: '14', path: '/14'	},
	{	text: '15', path: '/15'	},
	{	text: '16', path: '/16'	},
	{	text: '17', path: '/17'	},
	{	text: '18', path: '/18'	},
	{	text: '19', path: '/19'	},
	{	text: '20', path: '/20'	},
];

describe('Pagination component', () => {
	test('Matches the snapshot', () => {
		const component = renderer.create(<Pagination items={items} />).toJSON();
		expect(component).toMatchSnapshot();
	})
});
