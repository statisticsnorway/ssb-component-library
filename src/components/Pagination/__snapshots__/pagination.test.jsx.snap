// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component Matches the snapshot 1`] = `
<nav
  className="Pagination__PaginationWrapper-t7si0g-0 hEdbGl"
>
  <div
    className="Pagination__DirectionButton-t7si0g-1 fXdkih"
    onClick={[Function]}
  >
    <svg
      className="icon"
      fill="none"
      height={18}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={18}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="15 18 9 12 15 6"
      />
    </svg>
    <span>
      Forrige
    </span>
  </div>
  <div
    className="Pagination__NavButton-t7si0g-2 cGTCea"
    onClick={[Function]}
    selected={true}
  >
    1
  </div>
  <div
    className="Pagination__NavButton-t7si0g-2 khMUa"
    onClick={[Function]}
    selected={false}
  />
  <div
    className="Pagination__DottedIndicator-t7si0g-3 fTxHcT"
  >
    ...
  </div>
  <div
    className="Pagination__NavButton-t7si0g-2 khMUa"
    onClick={[Function]}
    selected={false}
  >
    20
  </div>
  <div
    className="Pagination__DirectionButton-t7si0g-1 fXdkih"
    onClick={[Function]}
  >
    <span>
      Neste
    </span>
    <svg
      className="icon"
      fill="none"
      height={18}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={18}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </div>
</nav>
`;
